src/field/generated_fieldops.cxx

```diff
@@ -0,0 +1,1668 @@
+// This file is autogenerated - see gen_fieldops.py
+#include <bout/mesh.hxx>
+#include <field2d.hxx>
+#include <field3d.hxx>
+#include <globals.hxx>
+#include <interpolation.hxx>
+
+// Do the actual multiplication of Field3D and Field3D
+void autogen_Field3D_Field3D_Field3D_multiplication(BoutReal *__restrict__ result,
+                                                    const BoutReal *__restrict__ lhs,
```

Comment:
Whilst probably not an issue as we only use read access for `lhs` and `rhs` presumably if we're doing `a = b * b` then we're not strictly avoiding aliasing. I may be misunderstanding what we're promising by using `__restrict__` and I'm not sure it can cause any problems anyway.

Reply:
I wasn't sure, but this is defined behaviour: http://en.cppreference.com/w/c/language/restrict
It's ok as long as neither `lhs` or `rhs` are modified.

I think this form is not the most efficient either. We probably want something like:

```cpp
for (int i = 0; i < len; ++i) {
    const BoutReal *__restrict__ lhs_value = lhs[i];
    const BoutReal *__restrict__ rhs_value = rhs[i];
    BoutReal *__restrict__ result_value = result[i];
    result_value = lhs_value + rhs_value;
}
```

See e.g. http://cellperformance.beyond3d.com/articles/2006/05/demystifying-the-restrict-keyword.html 
which talks about separating load, update, store operations (at least for gcc).

Reply:
I tried to use inline restrict pointer, but gcc didn't provide the performance, only after putting it in a separate function it worked.

further, the example does not work, as you only take a copy of the value, so there is no vectorization.

`a = b * b` is no issue, `a = a * b` is violating the promise, but I think it should still work.
Only times I had issues was something like `a = A * a` where `A` was a matrix, and some parts of `a` where written and then read again, but I cannot see that that would happen with a vector.

Another advantage of having separate function is, that we can put it in a separate file, and compile it with a C compiler, and use `restrict` rather then a gcc extension.